contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+    resource can hold additional metadata as well, for example:
+    * `nvidia.gpu.high.mem` is the name and metadata is memory greater than 'X' GB.
+    * `fast.nic` is the name and associated metadata is bandwidth greater than
+      'B' gbps.
+  * If I request a resource `nvidia.gpu.high.mem` for my pod, any 'nvidia-gpu'
+    type device which has memory greater than or equal to 'X' GB, should be able
+    to satisfy this request, independent of other device capabilities such as
+    'version' or 'nvlink locality' etc.
+  * Similarly, if I request a resource `fast.nic`, any nic device with speed
+    greater than 'B' gbps should be able to meet the request.
+  * I want a rich metadata selection interface where operators like 'Eq' for
+    'equals to', 'Lt' for 'less than', 'LtEq' for 'less than equal to', 'Gt' for
+    'greater than', 'GtEq' for 'greater than and equal to' and 'In' for
+    'a set of accepted values' are supported on the compute resource metadata.
+
+## Objectives
+
+1. Define and add support in the API for a new type, *Resource Class*.
+2. Add support for *Resource Class* in the scheduler.
+
+## Non Objectives
+1. Discovery, advertisement, allocation/deallocation of devices is expected to
+   be addressed by [device plugin proposal](https://github.com/kubernetes/community/pull/695/files)
+
+## Resource Class
+*Resource Class* is a new type, objects of which provide abstraction over
+[Devices](https://github.com/RenaudWasTaken/community/blob/a7762d8fa80b9a805dbaa7deb510e95128905148/contributors/design-proposals/device-plugin.md#resourcetype).
+A *Resource Class* object selects devices using `matchExpressions`, a list of
+(operator, key, value). A *Resource Class* object selects a device if at least
+one of the `matchExpressions` matches with device details. Within a matchExpression,
+all the (operator,key,value) are ANDed together to evaluate the result.
+
+*Resource Class* object is non-Namespaced kind of object and post created object
+is immutable.
+
+YAML example 1:
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia.high.mem
+spec:
+  resourceSelector:
+    - matchExpressions:
+        - key: "Kind"
+          operator: "In"
+          values:
+            - "nvidia-gpu"
+        - key: "memory"
+          operator: "GtEq"
+          values:
+            - "30G"
+```
+Above resource class will select all the nvidia-gpus which have memory greater
+than and equal to 30 GB.
+
+YAML example 2:
+```yaml
+kind: ResourceClass
+metadata:
+  name: hugepages-1gig
+spec:
+  resourceSelector:
+    - matchExpressions:
+        - key: "Kind"
+          operator: "In"
+          values:
+            - "huge-pages"
+        - key: "size"
+          operator: "GtEq"
+          values:
+            - "1G"
+```
+Above resource class will select all the hugepages with size greater than and
+equal to 1 GB.
+
+YAML example 3:
+```yaml
+kind: ResourceClass
+metadata:
+  name: fast.nic
```

Comment:
This is a key cross-sig deliverable.  We have discussed it at length in the RMWG and it was deferred because we wanted it to be led by sig-network (or more likely the new Network Plumbing WG).

Reply:
Is it possible that we start with a simple model to support high-performance nic with a combination of CNI plugin and device plugin? The CNI plugin takes care of network interface setup and management, and can be portable across different container orchestration systems. The device plugin can run as a sidecar container and take care of device initialization, health monitoring, and resource advertising. I know having device plugin act like a middleman between CNI and Kubelet may pose certain limitations. We can discuss them and see whether we may solve them by enriching the information passed between device plugin and Kubelet. However, at least for now, I hope the information passed between device plugin and Kubelet can stay at resource level (such as resource name and properties) and device level (such as device runtime configuration). This way, it is clear at the API level that Kubelet is the central place in charge of resource allocation and container runtime setup, which I feel perhaps will be easier to be extended to support future features like cross-resource affinity allocation and etc.

Reply:
I don't think the sig have enough resource to implement this case, I would like to see this design keep as simple as possible and hopefully get merged in 1.10 cycle.

Reply:
Ack, thanks for the info.

Reply:
@jiayingz technically it can also be done closer to CNI. To me it just looks like it will be a much cleaner interface to provide the network connectivity via the device manager. This will effectively formalize how components can add additional NICs to a pod. Today this is unspecified and everybody is pretty much using CNI in some way to achieve this.
The problem is that there are so many different assumptions by different projects. Eventually it even implies changing the CNI config of the host which is not necessarily desireable.

OTOH if we could use the device manager, then we know a way how to distribute such a plugin and how it would operate, which could eventually lead to more cooperation, and to less wild west.